File: offer/chaper4MoreEfficient/minStack/MinStack.java
package chaper4MoreEfficient.minStack;

import java.util.EmptyStackException;
import java.util.Stack;

/**
 * @author Weybn
 * @motto Rare in the world you're worth it.
 * @time 2020/2/17 15:40
 */
public class MinStack<T extends Comparable<T>> {
  Stack<T> normal;
  Stack<T> minStack;

  public MinStack() {
    this.normal = new Stack<>();
    this.minStack = new Stack<>();
  }

  public T pop() {
    if (normal.isEmpty()) {
      throw new EmptyStackException();
    }

    T item = normal.pop();
    if (minStack.peek().equals(item)) {
      minStack.pop();
    }

    return item;
  }

  public T getMin() {
    return minStack.peek();
  }

  public void push(T elem) {
    while (!minStack.isEmpty() && elem.compareTo(minStack.peek()) < 0) {
      minStack.pop();
    }

    normal.push(elem);
    minStack.push(elem);
  }
}

File: review_offer/MinStack.js
class MinStack {
  static peek(stack) {
    return stack[stack.length - 1];
  }

  constructor() {
    this.mainStack = [];
    this.minStack = [];
  }

  push(elem) {
    this.mainStack.push(elem);

    if (this.minStack.length === 0 || MinStack.peek(this.minStack) >= elem) {
      this.minStack.push(elem);
    }
  }

  pop() {
    let result = this.mainStack.pop();
    if (MinStack.peek(this.minStack) === result) {
      this.minStack.pop();
    }

    return result;
  }

  min() {
    return MinStack.peek(this.minStack);
  }
}

let stack = new MinStack();
stack.push(1);
stack.push(4);
stack.push(2);
stack.push(1);
stack.push(5);

console.log(stack.min());
stack.pop();
console.log(stack.min());
stack.pop();
stack.pop();
stack.pop();
console.log(stack.min());

